parquet-variant/src/variant.rs

```diff
@@ -0,0 +1,719 @@
+
+    // Note: only need the *value* buffer
+    String(&'v str),
+    ShortString(&'v str),
+
+    // need both metadata & value
+    Object(VariantObject<'m, 'v>),
+    Array(VariantArray<'m, 'v>),
+}
+
+impl<'m, 'v> Variant<'m, 'v> {
+    /// Parse the buffers and return the appropriate variant.
+    pub fn try_new(metadata: &'m VariantMetadata, value: &'v [u8]) -> Result<Self, ArrowError> {
+        let header = *first_byte_from_slice(value)?;
+        let new_self = match get_basic_type(header)? {
+            VariantBasicType::Primitive => match get_primitive_type(header)? {
+                VariantPrimitiveType::Null => Variant::Null,
+                VariantPrimitiveType::Int8 => Variant::Int8(decoder::decode_int8(value)?),
+                VariantPrimitiveType::BooleanTrue => Variant::BooleanTrue,
+                VariantPrimitiveType::BooleanFalse => Variant::BooleanFalse,
+                // TODO: Add types for the rest, once API is agreed upon
+                VariantPrimitiveType::String => {
+                    Variant::String(decoder::decode_long_string(value)?)
+                }
+            },
+            VariantBasicType::ShortString => {
+                Variant::ShortString(decoder::decode_short_string(value)?)
+            }
+            VariantBasicType::Object => Variant::Object(VariantObject { metadata, value }),
+            VariantBasicType::Array => Variant::Array(VariantArray { metadata, value }),
+        };
+        Ok(new_self)
+    }
+
+    pub fn as_null(&self) -> Option<()> {
+        matches!(self, Variant::Null).then_some(())
+    }
+
+    pub fn as_boolean(&self) -> Option<bool> {
+        match self {
+            Variant::BooleanTrue => Some(true),
+            Variant::BooleanFalse => Some(false),
+            _ => None,
+        }
+    }
+
+    pub fn as_string(&'v self) -> Option<&'v str> {
+        match self {
+            Variant::String(s) | Variant::ShortString(s) => Some(s),
+            _ => None,
+        }
+    }
+
+    pub fn as_int8(&self) -> Option<i8> {
+        match *self {
+            Variant::Int8(i) => Some(i),
+            // TODO: Add branches for type-widening/shortening when implemting rest of primitives for int
+            // Variant::Int16(i) => i.try_into().ok(),
+            // ...
+            _ => None,
+        }
+    }
+
+    pub fn metadata(&self) -> Option<&'m VariantMetadata> {
+        match self {
+            Variant::Object(VariantObject { metadata, .. })
+            | Variant::Array(VariantArray { metadata, .. }) => Some(*metadata),
+            _ => None,
+        }
+    }
+}
+
+impl<'m, 'v> From<i8> for Variant<'m, 'v> {
+    fn from(value: i8) -> Self {
+        Variant::Int8(value)
+    }
+}
+
+impl<'m, 'v> From<bool> for Variant<'m, 'v> {
+    fn from(value: bool) -> Self {
+        match value {
+            true => Variant::BooleanTrue,
+            false => Variant::BooleanFalse,
+        }
+    }
+}
+
+impl<'m, 'v> From<&'v str> for Variant<'m, 'v> {
```

Comment:
nit: I'm surprised clippy didn't warn about unnecessary named lifetimes:
```suggestion
impl<'v> From<&'v str> for Variant<'_, 'v> {
```